packages/queued-request-controller/src/QueuedRequestMiddleware.ts

```diff
@@ -0,0 +1,178 @@
+import { QueuedRequestControllerActionTypes } from './QueuedRequestController';
+
+const isConfirmationMethod = (method: string) => {
+  const confirmationMethods = [
+    'eth_sendTransaction',
+    'wallet_watchAsset',
+    'wallet_switchEthereumChain',
+    'eth_signTypedData_v4',
+  ];
+
+  return confirmationMethods.includes(method);
+};
+
+/**
+ * Creates a JSON-RPC middleware for handling queued requests. This middleware
+ * intercepts JSON-RPC requests, checks if they require queueing, and manages
+ * their execution based on the specified options.
+ *
+ * @param options - Configuration options.
+ * @param options.messenger - A controller messenger used for communication with various controllers.
+ * @param options.useRequestQueue - A function that determines if the request queue feature is enabled.
+ * @returns The JSON-RPC middleware that manages queued requests.
+ */
+export const createQueuedRequestMiddleware = ({
+  messenger,
+  useRequestQueue,
+}: {
+  messenger: ControllerMessenger<
+    | QueuedRequestControllerEnqueueRequestAction
+    | NetworkControllerGetStateAction
+    | NetworkControllerSetActiveNetworkAction
+    | NetworkControllerSetProviderTypeAction
+    | NetworkControllerGetNetworkClientByIdAction
+    | NetworkControllerFindNetworkClientIdByChainIdAction
+    | SelectedNetworkControllerSetNetworkClientIdForDomainAction
+    | AddApprovalRequest,
+    never
+  >;
+  useRequestQueue: () => boolean;
+}): JsonRpcMiddleware<JsonRpcParams, Json> => {
+  return createAsyncMiddleware(
+    async (
+      req: JsonRpcRequest & {
+        origin?: string;
+        networkClientId?: NetworkClientId;
+      },
+      res,
+      next,
+    ) => {
+      const { origin, networkClientId: networkClientIdForRequest } = req;
+
+      if (!origin) {
+        throw new Error("Request object is lacking an 'origin'");
+      }
+
+      if (!networkClientIdForRequest) {
+        throw new Error("Request object is lacking a 'networkClientId'");
+      }
+
+      // if the request queue feature is turned off, or this method is not a confirmation method
+      // do nothing
+      if (!useRequestQueue() || !isConfirmationMethod(req.method)) {
+        next();
+        return;
+      }
+
+      await messenger.call(
+        QueuedRequestControllerActionTypes.enqueueRequest,
+        async () => {
+          if (req.method === 'wallet_switchEthereumChain') {
+            return next();
+          }
+
+          const isBuiltIn = isNetworkType(networkClientIdForRequest);
+          let networkConfigurationForRequest;
+          if (isBuiltIn) {
+            const builtIn = BUILT_IN_NETWORKS[networkClientIdForRequest];
+            if (builtIn.chainId) {
+              // only the infura provided ones have chainid (rpc doesnt)
+              networkConfigurationForRequest = builtIn;
+            }
+          }
+
+          networkConfigurationForRequest ??= messenger.call(
```

Comment:
ooo you got fancy here 😄

Reply:
[nit] just because I don't see nullish assignment used very often it might be useful to add a comment. But not strictly necessary. Just took me a few seconds to grok

Reply:
yeah it was suggeted by one of the reviewers to use it. I think getting per dapp network working is far more important than using `??=` vs `networkConfigurationForRequest = networkConfigurationForRequest || xyz`. They both work, and we are yak shaving here by making these suggestions.

Reply:
yep that's why I put `[nit]`

Reply:
I think using `??` is important for reducing cognitive load for readers. If a `||` is used, I'll assume that's an intentional choice and spend time looking for the case where the variable could be falsy. `??` makes it clear that we're only guarding against the nullable case.